File: src/single_world/c/example/first.c
#include "libMPMDManager.h"
#include "message.h"

int main(int argc, char** argv) {
  MPI_Init(&argc, &argv);
  const MPMDManager* manager = Manager_Init(argv);
  int localRank = Manager_Local_rank(manager);

  char connectionName[10];
  ConnectionId nameId = {.name = connectionName};

  Message message;
  MPI_Datatype MessageType;
  getMessageType(&MessageType);

  if(localRank == 0) {
    strcpy(connectionName, "second");
    const MPI_Comm* secondComm = Manager_Comm_to(manager, nameId, NAME_ID);

    fillMessage(&message, "0|First  to 1|Second", 1021);
    MPI_Send(&message, 1, MessageType, 1, 0, *secondComm);
    printMessage("(Send) 0|First  -> 1|Second:", &message);
  } else if(localRank == 1) {
    strcpy(connectionName, "third");
    const MPI_Comm* thirdComm = Manager_Comm_to(manager, nameId, NAME_ID);

    MPI_Recv(&message, 1, MessageType, 0, 0, *thirdComm, MPI_STATUS_IGNORE);
    printMessage("(Recv) 1|First  <- 0|Third: ", &message);
  }

  Manager_Finalize(manager);
  MPI_Finalize();
}

File: src/single_world/python/example/first.py
from abstraction import MPMDManager

MPMDManager.init()
local_rank = MPMDManager.local_rank()

if(local_rank == 0):
  second_comm = MPMDManager.comm_to('second')
  transfer_object = {'text': '0|First  to 1|Second', 'value': 1021}
  second_comm.send(transfer_object, dest=1)
  
  print('(Send) 0|First  -> 1|Second: ', end='')
  print(transfer_object)
elif(local_rank == 1):
  third_comm = MPMDManager.comm_to('third')
  received_object = third_comm.recv(source=0)
  
  print('(Recv) 1|First  <- 0|Third: ', end=' ')
  print(received_object)

MPMDManager.finalize()
